### ISSUE TITLE
regl-renderer : improve performance of viewer

### ISSUE THREAD

**Issue by z3dev** [jscad/OpenJSCAD.org#812]
## Expected Behavior

Produce native buffers for use with WEB GL.

## Actual Behavior

Provided geometries (solids) are converted to entities for use in WEB GL, but the buffers are arrays of Number values. This results in an unnecessary conversion inside the regl library. (See line 122 of https://github.com/regl-project/regl/blob/master/lib/buffer.js that copies the array)

## Steps to Reproduce the Problem

  1. Load a large JSCAD design which take a long time to render

## Specifications

  - Version: V2+
  - Platform: WEB
  - Environment: browser

**Comment by hrgdavor** [comment]
@z3dev I have draft branch for this that works ok, but some major things came up that are far greater killer of performance...
more details in a new discussion #923 .

**Comment by z3dev** [comment]
First, I still think that rendering the geometry in the web worker is the the best direction to pursue. That allows everything to happen in the web worker, as well as rendering optimizations can be made.

Second, yeah. Geometry can be converted to buffers and posted back (maybe even transferable). That's why there are to/fromCompactBinary() functions. But that didn't turn out as planned, as general attributes cannot be passed as part of the buffers, so users are kind of 'out of luck' if they want to specify color, name, id, etc. as part of the design. It may not sound like a BIG issue but there are lots of use cases (many that we don't know) which will certainly use attributes.

Third, V2 has just been released within this year. And making the BIG changes to data structures means... yeah... V3. There's no way around this as there are some nice websites / applications being built with V2 modules, which was one of the major goals. V2 needs to continue on for a while longer.

So, what to do... i'm not 100% certain... but some playing around with a new web ui is certainly justified... and try some of these BIG improvements... maybe create a prototype

**Comment by hrgdavor** [comment]
postMessage to be fast does not need to have all data as native buffers

if you have a geometry: `{ color:[..], transforms:[..], polygons:[..]}` 
  - convert it to make a copy that uses native buffers 
  - new object will have all user properties but will replace `polygons` with native buffers `vertices`, `indices`, ``
  - fill `trensferables` a separate array that holds reference to each transferable(native array)
  - at the end postMessage([goems],[transferables])

```js
const prepareGeom = (geometry, transferables)=>{
  let out = {...geometry}
  delete geometry.polygons
  let vertices = ... do conversion
  let indices = ...  do conversion
  let colors = ... do for whn color per vertex is used
  transferables.push(vertices)
  transferables.push(indices)
  transferables.push(colors)
  return out
}

const doPostMessage= (geometries)=>{
  let transferables = []
  let out = geometries.map(geom=>prepareGeom(geom, transferables))
  postMessage(out,transferables)
}
```

Maybe it is best I make a prototype, and do stuff as I see fit(based on what I want to implement there) and when the prototype is good enough to replace current web, we can discuss if it is doable to still be a V2 improvement or be a V3.

All the changes I have in mind should be compatible with current v2 scripts unlike what happened in V1->V2, but having the integration(external use) be so different would maybe still warrant calling it V3.

I will create a branch, but it will not be a branch that can be a PR later, it will be discarde after the  features from the prototype are implemented properly. Not sure If i should do it in a separate repo even.

**Comment by hrgdavor** [comment]
I will create an initial prototype and a new discussion after that with first impressions. 

I was hoping to bring those new parts that I need for new prototype separately, but then you would not know the whole picture, why some choices are made etc.. 

I will say again: it will be messy, but I work faster that way, and can go faster through different ideas.

**Comment by z3dev** [comment]
Looking forward to it. I won’t be able to help much as I have some major difficult bugs to look at still. Also, there are some core changes to finish, and some IO enhancements. Work never ends!

**Comment by hrgdavor** [comment]
> First, I still think that rendering the geometry in the web worker is the the best direction to pursue. That allows everything to happen in the web worker, as well as rendering optimizations can be made.

I agree but the current setup where code-eval is in worker and render is on main thread will also need to be supported
 - to be able to use debugger and show geometry(from debug) at the same time
 - to support browsers without offlineCanvas (Safari and Firefox) sadly this is not that well supported as webgl2 :(

This can be done transparently, but will have to be communicated as an option for those seeking those last bits of performance.
